clang/lib/CConv/DeclRewriter.cpp

```diff
@@ -166,155 +166,179 @@ void DeclRewriter::rewriteParmVarDecl(ParmVarDeclReplacement *N) {
-      R.ReplaceText(TR, SRewrite);
-    } else {
-      // This can happen if SR is within a macro. If that is the case,
-      // maybe there is still something we can do because Decl refers
-      // to a non-macro line.
+void DeclRewriter::rewriteSingleDecl(DeclReplacement *N, RSet &ToRewrite) {
+  assert("Declaration is not a single declaration." && isSingleDeclaration(N));
+  // This is the easy case, we can rewrite it locally, at the declaration.
+  SourceRange TR = N->getDecl()->getSourceRange();
+  doDeclRewrite(TR, N);
+}
 
+void DeclRewriter::rewriteMultiDecl(DeclReplacement *N, RSet &ToRewrite) {
+  assert("Declaration is not a multi declaration." && !isSingleDeclaration(N));
+  // Rewriting is more difficult when there are multiple variables declared in a
+  // single statement. When this happens, we need to find all the declaration
+  // replacement for this statement and apply them at the same time. We also
+  // need to avoid rewriting any of these declarations twice by updating the
+  // Skip set to include the processed declarations.
+
+  // Step 1: get declaration replacement in the same statement
+  RSet RewritesForThisDecl(DComp(R.getSourceMgr()));
+  auto I = ToRewrite.find(N);
+  while (I != ToRewrite.end()) {
+    if (areDeclarationsOnSameLine(N, *I)) {
+      assert("Unexpected DeclReplacement kind." &&
+             (*I)->getKind() == N->getKind());
+      RewritesForThisDecl.insert(*I);
+    }
+    ++I;
+  }
 
-      SourceRange Possible(R.getSourceMgr().getExpansionLoc(TR.getBegin()),
-                           D->getEndLoc());
+  // Step 2: For each decl in the original, build up a new string. If the
+  //         original decl was re-written, write that out instead. Existing
+  //         initializers are preserved, any declarations that an initializer to
+  //         be valid checked-c are given one.
+  std::set<Decl *> SameLineDecls;
+  getDeclsOnSameLine(N, SameLineDecls);
+
+  bool IsFirst = true;
+  SourceLocation PrevEnd;
+  for (const auto &DL : SameLineDecls) {
+    // Find the declaration replacement object for the current declaration
+    DeclReplacement *SameLineReplacement;
+    bool Found = false;
+    for (const auto &NLT : RewritesForThisDecl)
+      if (NLT->getDecl() == DL) {
+        SameLineReplacement = NLT;
+        Found = true;
+        break;
+      }
 
-      if (canRewrite(R, Possible)) {
-        R.ReplaceText(Possible, SRewrite);
+    if (IsFirst) {
+      // Rewriting the first declaration is easy. Nothing should change if its
+      // type does not to be rewritten. When rewriting is required, it is
+      // essentially the same as the single declaration case.
+      IsFirst = false;
+      if (Found) {
+        SourceRange SR(DL->getBeginLoc(), DL->getEndLoc());
+        doDeclRewrite(SR, SameLineReplacement);
+      }
+    } else  {
+      // The subsequent decls are more complicated because we need to insert a
+      // type string even if the variables type hasn't changed.
```

Comment:
What happens if I have `int *p, *q` and only `p` becomes `_Ptr<int>` ? Does this code handle that, or is it somewhere else?

Reply:
(I'm guessing this code splits it into multiple `;` separated decls with initilaizers?)

Reply:
Yes, the `p` declaration is rewritten using the `doDeclRewrite` method that is also used for rewriting single declarations. Then, the comma is replaced with `;\n` to split the declaration before `q`.  The new type for `q` can then be inserted after this split.